### DISCUSSION TITLE
MapStruct in Spring application - is "spring" component model the right approach?

### DISCUSSION THREAD

**Discussion by maciejwalkowiak** [mapstruct/mapstruct#3208] [Q&A]
The reference [says](https://mapstruct.org/documentation/stable/reference/html/#using-dependency-injection): 
> If you’re working with a dependency injection framework such as [CDI](http://jcp.org/en/jsr/detail?id=346) (Contexts and Dependency Injection for JavaTM EE) or the [Spring Framework](http://www.springsource.org/spring-framework), it is recommended to obtain mapper objects via dependency injection and not via the Mappers (...)

I have some doubts if this is correct as I see potential harm in creating mappers as Spring beans. Perhaps I am missing something.

Lets consider following (oversimplified) example:

```java
public record Person(String firstName, String lastName, Address address) { }
public record Address(String street, String city) { }
// dtos
public record PersonDto(String firstName, String lastName, AddressDto address) { }
public record AddressDto(String streetName, String city) { }
```

My code without using MapStruct looks like this:

```java
@Service
public class PersonService {
    private final PersonClient personClient;

    public PersonService(PersonClient personClient) {
        this.personClient = personClient;
    }

    Person findById(Long id) {
        PersonDto dto = personClient.findById(id);
        return mapToPerson(dto);
    }

    private Person mapToPerson(PersonDto dto) {
        return new Person(dto.firstName(), dto.lastName(), new Address(dto.address().streetName(), dto.address().city()));
    }
}
```

The important part is that mapping is an implementation detail of the `PersonService` class. It could be also in the DTO itself or somewhere else, but it is hidden in the implementation of the `PersonService`.

Once I follow the recommended way and add mappers:

```java
@Mapper(uses = AddressMapper.class, injectionStrategy = InjectionStrategy.CONSTRUCTOR, componentModel = "spring")
public interface PersonMapper {
    Person mapToPerson(PersonDto dto);
}

@Mapper(injectionStrategy = InjectionStrategy.CONSTRUCTOR, componentModel = "spring")
public interface AddressMapper {
    @Mapping(source = "streetName", target = "street")
    Address mapToAddress(AddressDto dto);
}
```

... and modify the service class to use them:

```java
@Service
public class PersonService {
    private final PersonClient personClient;
    private final PersonMapper personMapper;

    public PersonService(PersonClient personClient, PersonMapper personMapper) {
        this.personClient = personClient;
        this.personMapper = personMapper;
    }

    Person findById(Long id) {
        PersonDto dto = personClient.findById(id);
        return personMapper.mapToPerson(dto);
    }
}
```

... the mapper suddenly becomes a dependency for a `PersonService` - which means when I instantiate it for example in tests, I must provide a mapper. 

I don't see why would I want to do it and what's the benefit of having mapper as a Spring bean. 

- there is no way I would want to provide an alternative implementation of the mapper (as introducing a mapper is just a refactoring of mapping that previously was an implementation detail of the `PersonService`)
- I should not mock the mapper in tests as mapping aspect is still an implementation detail
- it is very unlikely I would want to apply any AOP or any bean processing on mapper instances

Mappers being a Spring beans slow down Spring application context startup.
Mappers being a dependency drive developers to mock them in unit test (I know it's not the fault of the tool).

It sounds much more reasonable to me to keep stick to default component model and retrieve mappers from `Mappers` class:

```java
@Service
public class PersonService {
    private final PersonClient personClient;
    private final PersonMapper personMapper = Mappers.getMapper(PersonMapper.class);

    ...
```

Or even better, generate mappers with `CONSTRUCTOR` injection strategy **and** the `default` component model (currently not possible #2211).

```java
@Service
public class PersonService {
    private final PersonClient personClient;
    private final PersonMapper personMapper = new PersonMapperImpl(new AddressMapperImpl());

    ...
```

Since I haven't dug into MapStruct sources, perhaps I am missing something and there is a good reason to use Mappers as Spring beans. If that's the case I would appreciate if you could explain why.

**Comment by Chessray** [comment]
I'll just leave this here: [Mappers as Converters](https://mapstruct.org/documentation/spring-extensions/reference/html/#mapperAsConverter)

**Reply by maciejwalkowiak** [reply]
In the scenario mentioned in original post, why would I want to use mappers as converters?

**Reply by Chessray** [reply]
It's really the usual question: If your mappings are very simple, there's not too much benefit in using MapStruct at all - aside from not having to adjust your mapping code for simple changes like adding type-compatible fields with the same name to source and target.

Generally speaking, I just find it more consistent to use the component model matching the framework you're using anyway.

**Reply by rxynrg** [reply]
In my humble opinion, the use of 'mappers as converters' appears to be a more practical approach to organizing mapping in the world of Spring applications, in contrast to the proposed method involving explicit creation through a constructor or static call to `Mappers.getMapper()`. This is primarily due to the fact that such a technique breaches the widely adopted dependency injection in Spring applications. I believe that many people share this opinion.

**Reply by eriknyk** [reply]
Talking a little bit of a clean architecture the service class is in the domain layer and Datos should be in the presentation layer, so Datos and mapper should to be outside of your service class, then you can test your service class without any mapping or Datos dependencies, data presentation/transformations should occurs in the top layer of your app, by example only inject mappers and returns Datos in controllers classes

**Comment by RichMacDonald** [comment]
I abandoned using spring beans for Mappers. It got in the way of testing.

Mappers never change. You're not going to need to re-instantiate them for every unit test. They don't need to be mocked. They are truly singletons. The rest is decoration. The following is just fine and has minimal complexity.
````java
MyMapper INSTANCE = Mappers.getMapper(MyInstance.class);
````

Second issue I had with spring bean injection was that I have Mappers that refer to other Mappers. Good luck trying to inject a Mapper into another Mapper interface. Note: The @Mapper(uses=AnotherMapper.class) capability is insufficient because that mechanism does not pick up inheritance patterns.

**Comment by filiphr** [comment]
I already provided some answers to this question on the Twitter thread (https://twitter.com/maciejwalkowiak/status/1638853501726998532).

Basically nothing is forcing you to use the spring component model. If you want to use `Mapper#getMapper` in your Spring applications you are more than welcome to do that. 

The main reason why we suggest using the appropriate component model for your dependency injection framework is due to the fact that `Mapper#uses` allows you to use your own custom mappers / services. The classes defined in `Mapper#uses` do not have to be annotated with `@Mapper` themselves. If you do not use the spring component model MapStruct will try to instantiate this class / interface and this won't work.
